pkg/broker/broker.go

```diff
@@ -850,10 +859,194 @@ func (a AnsibleBroker) Unbind(
+		a.log.Debug("Error retrieving instance")
+		return nil, ErrorNotFound
+	}
+
+	// If no plan was specified we'll continue with the current plan
+	if req.PlanID == "" {
+		req.PlanID = (*si.Parameters)["_apb_plan_id"].(string)
+	}
+
+	// Retrieve requested spec
+	spec, err := a.dao.GetSpec(si.Spec.ID)
+	if err != nil {
+		// etcd return not found i.e. code 100
+		if client.IsKeyNotFound(err) {
+			return nil, ErrorNotFound
+		}
+		// otherwise unknown error bubble it up
+		return nil, err
+	}
+
+	// Make sure that we can find the plan we're updating from. This should probably never fail, but cover our tail.
+	var updateFromPlanKey = -1
+	for k, v := range spec.Plans {
+		if v.Name == (*si.Parameters)["_apb_plan_id"] {
+			updateFromPlanKey = k
+		}
+	}
+	if updateFromPlanKey == -1 {
+		a.log.Error("The plan %s, specified for updating from on instance %s, does not exist.", (*si.Parameters)["_apb_plan_id"], si.ID)
+		return nil, ErrorPlanNotFound
+	}
+	updateFromPlan := spec.Plans[updateFromPlanKey]
+
+	// Make sure we can find the plan we're updating to.
+	var updateToPlanKey = -1
+	for k, v := range spec.Plans {
+		if v.Name == req.PlanID {
+			updateToPlanKey = k
+		}
+	}
+	if updateToPlanKey == -1 {
+		a.log.Error("The plan %s, specified for updating to on instance %s, does not exist.", req.PlanID, si.ID)
+		return nil, ErrorPlanNotFound
+	}
+	updateToPlan := spec.Plans[updateFromPlanKey]
+
+	// Make sure that the current plan supports updating to the requested plan if a plan update was requested.
+	if req.PlanID != (*si.Parameters)["_apb_plan_id"] {
+		var updatable = false
+		for _, v := range updateFromPlan.UpdatesTo {
+			if v == req.PlanID {
+				updatable = true
+				(*si.Parameters)["_apb_plan_id"] = req.PlanID
+			}
+		}
+		if updatable == false {
+			a.log.Error("The current plan, %s, cannot be updated to the requested plan, %s.", (*si.Parameters)["_apb_plan_id"], req.PlanID)
+			return nil, ErrorPlanUpdateNotPossible
+		}
+	}
+
+	// Make sure that parameters requested for update exist on the new plan and that they are updatable.
+	for k, v := range req.Parameters {
+		var paramUpdated = false
+		for _, key := range updateToPlan.Parameters {
+			if key.Name == k {
+				if key.Updatable == false {
+					a.log.Error("Tried to update non-updatable parameter, %s, on instance %s.", k, si.ID)
+					return nil, ErrorParameterNotUpdatable
+				}
+				(*si.Parameters)[k] = v
+				paramUpdated = true
+			}
+		}
+		if paramUpdated == false {
+			a.log.Error("Parameter %s, requested for update on instance %s, does not exist.", k, si.ID)
+			return nil, ErrorParameterNotFound
+		}
+	}
+
+	// We're ready to provision so save
+	if err = a.dao.SetServiceInstance(instanceUUID.String(), si); err != nil {
+		return nil, err
+	}
+
+	var token string
+
+	if async {
+		a.log.Info("ASYNC update in progress")
+		// asyncronously provision and return the token for the lastoperation
+		pjob := NewProvisionJob("update", si, a.clusterConfig, a.log)
```

Comment:
We probably want a strongly typed `UpdateJob` suite, even if they do the same thing as provison. If the latter is true, we should refactor to account for all the shared stuff and better delegate specifics. This is kiind of a legacy thing, I'm happy  to help untangle this in another PR.

Reply:
Do we think this is a blocking issue for merging this PR, or should this get logged as an issue and submitted in a separate PR?

Reply:
@eriknelson I think a separate PR should be fine IMO